### ISSUE TITLE
Script in container is not exectued as root

### ISSUE THREAD

**Issue by gpad** [microsoft/azure-pipelines-agent#2619]
## Having issue with YAML?
Please log an issue at [Azure-Pipelines-YAML](https://github.com/Microsoft/azure-pipelines-yaml). Over there we discuss YAML templates, samples for Azure Pipelines, and designs for upcoming YAML features. Also a place for the community to share best practices, ideas, and so on. File suggestions and issues here if they're specific to YAML pipelines.

## Having issue with Tasks?
Log an issue at [Azure-Pipelines-Tasks](https://github.com/Microsoft/azure-pipelines-tasks). It contains all of the in-box tasks we ship with Azure-Pipelines/VSTS/TFS. If you're having issues with tasks in Build/Release jobs (e.g. unreasonable task failure) please log an issue there.

## Having issue with software on Hosted Agent?
Log an issue at [Hosted Agent Image Repository](https://github.com/Microsoft/vsts-image-generation). It contains the VM image used in the Azure Pipelines Hosted Agent Pool. If you're having Build/Release failures that seems like they are related to software installed on the Hosted Agent (e.g. the `dotnet` SDK is missing or the Azure SDK is not on the latest version) please log an issue there.

## Having generic issue with Azure-Pipelines/VSTS/TFS?
Please report it on [Developer Community](https://developercommunity.visualstudio.com/spaces/21/index.html)

## Have you tried troubleshooting?
[Troubleshooting doc](https://www.visualstudio.com/en-us/docs/build/troubleshooting)

## Agent Version and Platform
Version of your agent? 2.144.0/2.144.1/...

Pool: Azure Pipelines
Image: Ubuntu16
Agent: Hosted Agent
Started: Yesterday at 12:18 PM
Duration: 41s

Starting: Initialize job
Agent name: 'Hosted Agent'
Agent machine name: 'fv-az592'
Current agent version: '2.160.1'
Current image version: '20191025.1'
Agent running as: 'vsts'

## Azure DevOps Type and Version
dev.azure.com 

If dev.azure.com, what is your organization name? https://dev.azure.com/manpremo

## What's not working?
Please include error messages and screenshots.

Running a container image like for example node:10 I should be root and I should have admin right but I'm not root (I'm vsts_azpcontainer) and I don't have admin right so for example I can't update npm why?

## Agent and Worker's Diagnostic Logs

```
vsts_azpcontainer
npm WARN checkPermissions Missing write access to /usr/local/lib/node_modules/npm/node_modules/chownr
npm WARN checkPermissions Missing write access to /usr/local/lib/node_modules/npm/node_modules/graceful-fs
npm WARN checkPermissions Missing write access to /usr/local/lib/node_modules/npm/node_modules/hosted-git-info
npm WARN checkPermissions Missing write access to /usr/local/lib/node_modules/npm/node_modules/https-proxy-agent
npm WARN checkPermissions Missing write access to /usr/local/lib/node_modules/npm/node_modules/ignore-walk
npm WARN checkPermissions Missing write access to /usr/local/lib/node_modules/npm/node_modules/make-fetch-happen
npm WARN checkPermissions Missing write access to /usr/local/lib/node_modules/npm/node_modules/minipass/node_modules/yallist
npm WARN checkPermissions Missing write acc
```

**Comment by jtpetty** [comment]
@gpad - can you run the script as sudo? will that solve your issue?

For context, here is some discussion on this topic: #2043

**Comment by gpad** [comment]
@jtpetty I try to run the command as `sudo` but `sudo` is not installed on the current image ... I have checked the issue but it's almost the same problem. With standard image it's impossible to use `sudo`

There isn't a solution ... (mount a volume with sudo installed is not solution...)

*TL; DR*

I really don't understand this policy of _change the user_ and also their rights, of Azure Pipeline. In this way a lot of benefits of docker are vanished. You run a script in the same image locally and in Azure Pipelne and the context is different (locally is root with admin rights) on Azure it's a different user without admin right ... why ?!?!?.

**Comment by jtpetty** [comment]
@gpad - I was not around when the decision was made to run commands inside the container as a non-root user. Maybe @vtbassmatt can point us to some docs as to why.

**Comment by gpad** [comment]
This is a very big issue because you can't install package, update something inside docker and made all the ecosystem almost impossible to manage ...

**Comment by vtbassmatt** [comment]
I know we've discussed this in another thread, which I'm not finding off-hand. One fundamental challenge for us is file permissions of things touched by / created in the container. Artifact upload and several other asynchronous operations happen host-side, so the agent's userId has to be able to read the working directory. We ensure that by making the userId inside the container match the userId outside the container. (We can't just run them inside the container, either, because they depend on .NET Core. I suspect users would be equally unhappy if we required .NET Core inside every container!)

This is unusual for containers, I agree. The alternatives we evaluated, including using `docker copy` to replicate the files out or requiring the agent to be run as root on the host, all were worse.

**Comment by gpad** [comment]
Thanks @vtbassmatt  the real problem is that, in this way, inside the container is almost impossible execute some operation as root. It's impossible install some package, it's impossible write on some folder ... This choice made all the service almost fragile because lost one of the great feature docker reproducibility. Some script executed inside a docker works for me and not for azure pipeline

**Comment by vtbassmatt** [comment]
When this decision was made (before my time, to be clear :grin:) the belief was that build job containers would be immutable - that customers would be using it to snapshot tools/SDKs/etc required in a build. That's how containers are used for deployments, for instance. With the benefit of hindsight, that's not always how build containers are being used. Some of the follow-on decisions make for a painful experience if you intend to do things like install software in the container at runtime.

We're still considering options for fixing this (without breaking the satisfied majority of customers).

**Comment by prakashguru** [comment]
@vtbassmatt I had a similar issue. 

I have bunch of dotnet tools "sonarscanner" and "trx2junit" installed inside the container during the container build. i.e. The container dockerfile has commands to install those as dotnet global tools. https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-tool-install#description 

.NET global tools are usually installed in "$HOME/.dotnet/tools" which "/root/.dotnet/tools" during docker build. But when I use the azure devops container jobs, The commands are unable to find the tool "Could not execute because the specified command or file was not found.". 

Because the container jobs are runs in different user and the global tool was missing. 

Do you have any solution / suggestion?

**Comment by vtbassmatt** [comment]
@prakashguru as a workaround, it looks like you can [install global tools in a custom location](https://docs.microsoft.com/en-us/dotnet/core/tools/global-tools#install-a-global-tool-in-a-custom-location). Then make sure that custom location is available on `PATH`.

**Comment by dipanshuchaubey** [comment]
I have found a way to run scripts in container as root.

Step 1: Add your container resource in azure pipeline 
```yaml
resources:
  containers:
    - container: test
      image: image-name:latest
      options: "--name sample -v /usr/bin/docker:/tmp/docker:ro"
```

Step 2: In jobs, add a step where you can install sudo in the container, This will help you run commands as root
```yaml
pool:
      vmImage: "ubuntu-20.04"

    container: test

    steps:
      - script: |
          /tmp/docker exec -t -u root sample mv /etc/sudoers /etc/sudoers.bak
          /tmp/docker exec -t -u root sample apt install -y sudo
          /tmp/docker exec -t -u root sample mv /etc/sudoers.bak /etc/sudoers
          sudo chmod 777 /home/glue_user /tmp/spark-events/
```

Now, you have installed sudo in your container.

**Comment by fshstk** [comment]
Sorry for digging this one up but... maybe I'm missing something fundamental, but I'm surprised this isn't a much larger issue for more people.

One of the key selling points of Dockerized CI workflows is having perfectly reproducible environments. This includes installing dependencies and setting env variables, both of which are generally done for the specific user the container is designed to run as. (root, or some other user defined in the Dockerfile, whatever.)

Bearing this in mind, it seems to me that forcing CI jobs to run as a new user not only undermines, but flat out defeats the entire purpose of having a containerized workflow in the first place.

To give a more specific example, I am currently struggling with setting up CI builds for some embedded projects. We have Docker images with installed toolchains and SDKs. These SDKs rely on a specific user being logged in: certain ENV variables are set by our Dockerfile, a bash script needs to be sourced (which is added to ~/.bashrc by our Dockerfile), and the SDK directory needs to be owned by the current user, or nothing will work.

Now, there may be hacky workarounds for all three of these problems, but tbh having reproducible environments without hacky workarounds was the whole reason our dev team decided to switch to containers to begin with, and this issue completely obliterates that benefit as far as I can tell. Anyone with access to our Docker image can run `docker run -it --rm -v [...]/repo:/repo <image>` and then build any of our repositories with no additional effort, yet this is currently straight up impossible with Azure Pipelines.

Please enlighten me if I am missing some key point here...

**Comment by trondhindenes** [comment]
this is insane.
